pip install hangs with interactive setup.py setups

Try:

```
pip install egenix-pyopenssl
```

with Python 2.7 to see what I mean.

The setup.py in the packages asks for confirmation and pip no longer seems to support this (it did with earlier pip versions). Instead, the pip process simply hangs.

```console
$ python setup.py -qqqqqqqq install -qqqqqqq --single-version-externally-managed --record /dev/null
package init file 'warehouse/migrations/__init__.py' not found (or not a regular file)
warning: no previously-included files found matching 'requirements.txt'
warning: no previously-included files found matching 'requirements-dev.txt'
warning: no previously-included files found matching 'requirements-deploy.txt'
warning: no previously-included files found matching 'babel.cfg'
warning: no previously-included files found matching 'Makefile'
warning: no previously-included files found matching 'Gulpfile.js'
warning: no previously-included files found matching 'package.json'
warning: no previously-included files found matching '.dockerignore'
warning: no previously-included files found matching 'Dockerfile'
warning: no previously-included files found matching 'docker-compose.yml'
warning: no previously-included files found matching 'Procfile'
warning: no previously-included files found matching 'runtime.txt'
warning: no previously-included files found matching 'uwsgi.ini'
no previously-included directories found matching 'bin'
no previously-included directories found matching 'vcl'
no previously-included directories found matching '.travis'
no previously-included directories found matching 'dev'
no previously-included directories found matching 'docs/_build'
no previously-included directories found matching 'tasks'
```

That's still a page full of stuff that nobody cares about and is completely useless.

Here's what the output looks like if you have a file in it that for 2.x when you're installing into 3.x (or the other way too):

```console
$ python setup.py -qqqqqqqq install -qqqqqqq --single-version-externally-managed --record /dev/null
package init file 'warehouse/migrations/__init__.py' not found (or not a regular file)
warning: no previously-included files found matching 'requirements.txt'
warning: no previously-included files found matching 'requirements-dev.txt'
warning: no previously-included files found matching 'requirements-deploy.txt'
warning: no previously-included files found matching 'babel.cfg'
warning: no previously-included files found matching 'Makefile'
warning: no previously-included files found matching 'Gulpfile.js'
warning: no previously-included files found matching 'package.json'
warning: no previously-included files found matching '.dockerignore'
warning: no previously-included files found matching 'Dockerfile'
warning: no previously-included files found matching 'docker-compose.yml'
warning: no previously-included files found matching 'Procfile'
warning: no previously-included files found matching 'runtime.txt'
warning: no previously-included files found matching 'uwsgi.ini'
no previously-included directories found matching 'bin'
no previously-included directories found matching 'vcl'
no previously-included directories found matching '.travis'
no previously-included directories found matching 'dev'
no previously-included directories found matching 'docs/_build'
no previously-included directories found matching 'tasks'
  File "/Users/dstufft/.virtualenvs/tmp-8474a91e5e17e69/lib/python3.5/site-packages/warehouse/invalid_syntax.py", line 1
    raise Exception, "Bar"
                   ^
SyntaxError: invalid syntax
```

This install was completely successful, the error there is only an error during byte compile and can safely be ignored... but how is someone supposed to know that from the output? This was a constant source of confusion for people.

Even for the Pillow case, it does mean that the informational stuff shows up on the first page but it still shows stuff that is unlikely to be useful for anyone but the maintainers of Pillow:

```console
$ python setup.py -qqqqqqqq install -qqqqqqq --single-version-externally-managed --record /dev/null
--- using frameworks at /System/Library/Frameworks
Building using 4 processes
_imaging.c:488:7: warning: variable 'r' is used uninitialized whenever 'if' condition is false [-Wsometimes-uninitialized]
                if (PyLong_Check(color)) {
                    ^~~~~~~~~~~~~~~~~~~
/Users/dstufft/.pyenv/versions/3.5.0/include/python3.5m/longobject.h:15:9: note: expanded from macro 'PyLong_Check'
        PyType_FastSubclass(Py_TYPE(op), Py_TPFLAGS_LONG_SUBCLASS)
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/dstufft/.pyenv/versions/3.5.0/include/python3.5m/object.h:666:35: note: expanded from macro 'PyType_FastSubclass'
#define PyType_FastSubclass(t,f)  PyType_HasFeature(t,f)
                                  ^~~~~~~~~~~~~~~~~~~~~~
/Users/dstufft/.pyenv/versions/3.5.0/include/python3.5m/object.h:664:33: note: expanded from macro 'PyType_HasFeature'
#define PyType_HasFeature(t,f)  (((t)->tp_flags & (f)) != 0)
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
_imaging.c:498:7: note: uninitialized use occurs here
                if (r == -1 && PyErr_Occurred())
                    ^
_imaging.c:488:3: note: remove the 'if' if its condition is always true
                if (PyLong_Check(color)) {
                ^~~~~~~~~~~~~~~~~~~~~~~~~
_imaging.c:477:10: note: initialize the variable 'r' to silence this warning
    int r, g, b, a;
         ^
          = 0
1 warning generated.
encode.c:57:43: warning: comparison of constant 0 with boolean expression is always false
      [-Wtautological-constant-out-of-range-compare]
    if(!PyType_Ready(&ImagingEncoderType) < 0)
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~
1 warning generated.
libImaging/JpegEncode.c:233:37: warning: comparison of integers of different signs: 'int' and 'size_t' (aka 'unsigned long')
      [-Wsign-compare]
        if (context->rawExifLen + 5 >  context->destination.pub.free_in_buffer){
            ~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
libImaging/QuantOctree.c:188:20: warning: comparison of integers of different signs: 'unsigned int' and 'int' [-Wsign-compare]
   if (cube->rBits > rBits) {
       ~~~~~~~~~~~ ^ ~~~~~
libImaging/QuantOctree.c:195:20: warning: comparison of integers of different signs: 'unsigned int' and 'int' [-Wsign-compare]
   if (cube->gBits > gBits) {
       ~~~~~~~~~~~ ^ ~~~~~
libImaging/QuantOctree.c:202:20: warning: comparison of integers of different signs: 'unsigned int' and 'int' [-Wsign-compare]
   if (cube->bBits > bBits) {
       ~~~~~~~~~~~ ^ ~~~~~
libImaging/QuantOctree.c:209:20: warning: comparison of integers of different signs: 'unsigned int' and 'int' [-Wsign-compare]
   if (cube->aBits > aBits) {
       ~~~~~~~~~~~ ^ ~~~~~
4 warnings generated.
Building using 4 processes
_imagingft.c:400:27: warning: comparison of integers of different signs: 'int' and 'unsigned int' [-Wsign-compare]
            for (y = 0; y < glyph->bitmap.rows; y++) {
                        ~ ^ ~~~~~~~~~~~~~~~~~~
_imagingft.c:419:27: warning: comparison of integers of different signs: 'int' and 'unsigned int' [-Wsign-compare]
            for (y = 0; y < glyph->bitmap.rows; y++) {
                        ~ ^ ~~~~~~~~~~~~~~~~~~
2 warnings generated.
Building using 4 processes
Building using 4 processes
clangclang: : warning: warning: -framework Tcl: 'linker' input unused-framework Tcl: 'linker' input unused

clangclang: : warningwarning: : -framework Tk: 'linker' input unused-framework Tk: 'linker' input unused

Building using 4 processes
Building using 4 processes
--------------------------------------------------------------------
PIL SETUP SUMMARY
--------------------------------------------------------------------
version      Pillow 3.0.0
platform     darwin 3.5.0 (default, Oct  8 2015, 21:08:25)
             [GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.72)]
--------------------------------------------------------------------
--- TKINTER support available
--- JPEG support available
*** OPENJPEG (JPEG2000) support not available
--- ZLIB (PNG/ZIP) support available
--- LIBTIFF support available
--- FREETYPE2 support available
*** LITTLECMS2 support not available
--- WEBP support available
--- WEBPMUX support available
--------------------------------------------------------------------
To add a missing option, make sure you have the required
library, and set the corresponding ROOT variable in the
setup.py script.

To check the build, run the selftest.py script.
```

It's easy to look at something like that Pillow example and just sort of ignore the fact that there is a good chunk of stuff there that most people just won't care about or even the "simple" example that has a page full of stuff. However, when you're installing a number of projects like you will with any moderately complex project each page of output produced by these things add up and it's really easy for even minimal output to end up with 10-30 pages of output where the vast bulk of it simply isn't useful at all or worse, is outright confusing.

If we were a GUI we could hide the build log under a collapsed thing for each thing we installed which would make it way easier, but we're not. Reducing the amount of pointless output is important to a lot of our users and since showing it only helps if you're building the thing right now, not if you're re-using the wheel you built a month ago, I think this is better solved by libraries providing more insight against the installed item than by increasing the verbosity of pip by default.

That being said, it's possible that pip *could* store the output of the ``setup.py`` invocation in the installation metadata... but it's not entirely trivial to do since in the wheel build case we won't have that until after the wheel is built. So we'd need to then open the wheel and add a file to it. This would ruin the ability for the wheel file to be deterministic though (which isn't currently possible, but it's an oft requested feature).

On 02.11.2015 16:38, Donald Stufft wrote:
> ```console
> $ python setup.py -qqqqqqqq install -qqqqqqq --single-version-externally-managed --record /dev/null
> package init file 'warehouse/migrations/__init__.py' not found (or not a regular file)
> warning: no previously-included files found matching 'requirements.txt'
> warning: no previously-included files found matching 'requirements-dev.txt'
> warning: no previously-included files found matching 'requirements-deploy.txt'
> warning: no previously-included files found matching 'babel.cfg'
> warning: no previously-included files found matching 'Makefile'
> warning: no previously-included files found matching 'Gulpfile.js'
> warning: no previously-included files found matching 'package.json'
> warning: no previously-included files found matching '.dockerignore'
> warning: no previously-included files found matching 'Dockerfile'
> warning: no previously-included files found matching 'docker-compose.yml'
> warning: no previously-included files found matching 'Procfile'
> warning: no previously-included files found matching 'runtime.txt'
> warning: no previously-included files found matching 'uwsgi.ini'
> no previously-included directories found matching 'bin'
> no previously-included directories found matching 'vcl'
> no previously-included directories found matching '.travis'
> no previously-included directories found matching 'dev'
> no previously-included directories found matching 'docs/_build'
> no previously-included directories found matching 'tasks'
> ```
> 
> That's still a page full of stuff that nobody cares about and is completely useless.

Is it ?

At least some of those warnings and errors point to problems
with your MANIFEST.in file (or the lack of having such a file
and distutils assuming some defaults).

> Here's what the output looks like if you have a file in it that for 2.x when you're installing into 3.x (or the other way too):
> 
> ```console
> $ python setup.py -qqqqqqqq install -qqqqqqq --single-version-externally-managed --record /dev/null
> package init file 'warehouse/migrations/__init__.py' not found (or not a regular file)
> warning: no previously-included files found matching 'requirements.txt'
> warning: no previously-included files found matching 'requirements-dev.txt'
> warning: no previously-included files found matching 'requirements-deploy.txt'
> warning: no previously-included files found matching 'babel.cfg'
> warning: no previously-included files found matching 'Makefile'
> warning: no previously-included files found matching 'Gulpfile.js'
> warning: no previously-included files found matching 'package.json'
> warning: no previously-included files found matching '.dockerignore'
> warning: no previously-included files found matching 'Dockerfile'
> warning: no previously-included files found matching 'docker-compose.yml'
> warning: no previously-included files found matching 'Procfile'
> warning: no previously-included files found matching 'runtime.txt'
> warning: no previously-included files found matching 'uwsgi.ini'
> no previously-included directories found matching 'bin'
> no previously-included directories found matching 'vcl'
> no previously-included directories found matching '.travis'
> no previously-included directories found matching 'dev'
> no previously-included directories found matching 'docs/_build'
> no previously-included directories found matching 'tasks'
>   File "/Users/dstufft/.virtualenvs/tmp-8474a91e5e17e69/lib/python3.5/site-packages/warehouse/invalid_syntax.py", line 1
>     raise Exception, "Bar"
>                    ^
> SyntaxError: invalid syntax
> ```
> 
> This install was completely successful, the error there is only an error during byte compile and can safely be ignored... but how is someone supposed to know that from the output? This was a constant source of confusion for people.

By looking at it and seeing that Python 2 syntax is not being accepted
by Python 3 ?!

The above may also well be a genuine error in the application,
e.g. you may be trying to install Python 2 software with
Python 3.

> Even for the Pillow case, it does mean that the informational stuff shows up on the first page but it still shows stuff that is unlikely to be useful for anyone but the maintainers of Pillow:
> 
> ```console
> $ python setup.py -qqqqqqqq install -qqqqqqq --single-version-externally-managed --record /dev/null

BTW: You only need a single -q, since this sets .verbosity to 0.

> --- using frameworks at /System/Library/Frameworks

This is useful information.

> Building using 4 processes
> _imaging.c:488:7: warning: variable 'r' is used uninitialized whenever 'if' condition is false [-Wsometimes-uninitialized]
>                 if (PyLong_Check(color)) {
>                     ^~~~~~~~~~~~~~~~~~~
> /Users/dstufft/.pyenv/versions/3.5.0/include/python3.5m/longobject.h:15:9: note: expanded from macro 'PyLong_Check'
>         PyType_FastSubclass(Py_TYPE(op), Py_TPFLAGS_LONG_SUBCLASS)
>         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /Users/dstufft/.pyenv/versions/3.5.0/include/python3.5m/object.h:666:35: note: expanded from macro 'PyType_FastSubclass'
> #define PyType_FastSubclass(t,f)  PyType_HasFeature(t,f)
>                                   ^~~~~~~~~~~~~~~~~~~~~~
> /Users/dstufft/.pyenv/versions/3.5.0/include/python3.5m/object.h:664:33: note: expanded from macro 'PyType_HasFeature'
> #define PyType_HasFeature(t,f)  (((t)->tp_flags & (f)) != 0)
>                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
> _imaging.c:498:7: note: uninitialized use occurs here
>                 if (r == -1 && PyErr_Occurred())
>                     ^
> _imaging.c:488:3: note: remove the 'if' if its condition is always true
>                 if (PyLong_Check(color)) {
>                 ^~~~~~~~~~~~~~~~~~~~~~~~~
> _imaging.c:477:10: note: initialize the variable 'r' to silence this warning
>     int r, g, b, a;
>          ^
>           = 0
> 1 warning generated.
> encode.c:57:43: warning: comparison of constant 0 with boolean expression is always false
>       [-Wtautological-constant-out-of-range-compare]
>     if(!PyType_Ready(&ImagingEncoderType) < 0)
>        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~
> 1 warning generated.
> libImaging/JpegEncode.c:233:37: warning: comparison of integers of different signs: 'int' and 'size_t' (aka 'unsigned long')
>       [-Wsign-compare]
>         if (context->rawExifLen + 5 >  context->destination.pub.free_in_buffer){
>             ~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 1 warning generated.
> libImaging/QuantOctree.c:188:20: warning: comparison of integers of different signs: 'unsigned int' and 'int' [-Wsign-compare]
>    if (cube->rBits > rBits) {
>        ~~~~~~~~~~~ ^ ~~~~~
> libImaging/QuantOctree.c:195:20: warning: comparison of integers of different signs: 'unsigned int' and 'int' [-Wsign-compare]
>    if (cube->gBits > gBits) {
>        ~~~~~~~~~~~ ^ ~~~~~
> libImaging/QuantOctree.c:202:20: warning: comparison of integers of different signs: 'unsigned int' and 'int' [-Wsign-compare]
>    if (cube->bBits > bBits) {
>        ~~~~~~~~~~~ ^ ~~~~~
> libImaging/QuantOctree.c:209:20: warning: comparison of integers of different signs: 'unsigned int' and 'int' [-Wsign-compare]
>    if (cube->aBits > aBits) {
>        ~~~~~~~~~~~ ^ ~~~~~
> 4 warnings generated.
> Building using 4 processes
> _imagingft.c:400:27: warning: comparison of integers of different signs: 'int' and 'unsigned int' [-Wsign-compare]
>             for (y = 0; y < glyph->bitmap.rows; y++) {
>                         ~ ^ ~~~~~~~~~~~~~~~~~~
> _imagingft.c:419:27: warning: comparison of integers of different signs: 'int' and 'unsigned int' [-Wsign-compare]
>             for (y = 0; y < glyph->bitmap.rows; y++) {
>                         ~ ^ ~~~~~~~~~~~~~~~~~~
> 2 warnings generated.

These warnings may well hint at problems when running the
application and could cause a crash or unexpected output.

Maintainers should be made aware of these warnings.

> Building using 4 processes
> Building using 4 processes
> clangclang: : warning: warning: -framework Tcl: 'linker' input unused-framework Tcl: 'linker' input unused
> 
> clangclang: : warningwarning: : -framework Tk: 'linker' input unused-framework Tk: 'linker' input unused


Again, this is something to report to the maintainers. That
way they can document or fix the warning in the setup.py output.

> Building using 4 processes
> Building using 4 processes
> --------------------------------------------------------------------
> PIL SETUP SUMMARY
> --------------------------------------------------------------------
> version      Pillow 3.0.0
> platform     darwin 3.5.0 (default, Oct  8 2015, 21:08:25)
>              [GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.72)]
> --------------------------------------------------------------------
> --- TKINTER support available
> --- JPEG support available
> *** OPENJPEG (JPEG2000) support not available
> --- ZLIB (PNG/ZIP) support available
> --- LIBTIFF support available
> --- FREETYPE2 support available
> *** LITTLECMS2 support not available
> --- WEBP support available
> --- WEBPMUX support available
> --------------------------------------------------------------------
> To add a missing option, make sure you have the required
> library, and set the corresponding ROOT variable in the
> setup.py script.
> 
> To check the build, run the selftest.py script.
> ```
> 
> It's easy to look at something like that Pillow example and just sort of ignore the fact that there is a good chunk of stuff there that most people just won't care about or even the "simple" example that has a page full of stuff. However, when you're installing a number of projects like you will with any moderately complex project each page of output produced by these things add up and it's really easy for even minimal output to end up with 10-30 pages of output where the vast bulk of it simply isn't useful at all or worse, is outright confusing.
> 
> If we were a GUI we could hide the build log under a collapsed thing for each thing we installed which would make it way easier, but we're not. Reducing the amount of pointless output is important to a lot of our users and since showing it only helps if you're building the thing right now, not if you're re-using the wheel you built a month ago, I think this is better solved by libraries providing more insight against the installed item than by increasing the verbosity of pip by default.

I think the answer to all this needs to be more nuanced. On/off is
not the right way to approach this. If a build script needs
to report important information to the user, this should be
possible, either by using special syntax for the output or by
signaling it in some other way.

The full log then provides the necessary audit information for
later debugging or finding out why certain things failed to
build.

In the above -q examples, (most, if not all) output to stdout is
silenced. What you are seeing on the console is mostly stderr
output generated by the compiler and warnings issued using
log.warn() from inside distutils.

> That being said, it's possible that pip *could* store the output of the ``setup.py`` invocation in the installation metadata... but it's not entirely trivial to do since in the wheel build case we won't have that until after the wheel is built. So we'd need to then open the wheel and add a file to it. This would ruin the ability for the wheel file to be deterministic though (which isn't currently possible, but it's an oft requested feature).

Why not store the log alongside the wheel file in the
cache and have pip output the log using e.g.
"pip audit somepackage" ?

-- 
Marc-Andre Lemburg
http://www.malemburg.com/

--- a/news/2732.bugfix
+++ b/news/2732.bugfix
@@ -1,0 +1,1 @@
+Interactive setup.py files will no longer hang indefinitely.

--- a/news/4982.bugfix
+++ b/news/4982.bugfix
@@ -1,0 +1,1 @@
+Interactive setup.py files will no longer hang indefinitely.

--- a/src/pip/_internal/utils/misc.py
+++ b/src/pip/_internal/utils/misc.py
@@ -678,9 +678,10 @@ def call_subprocess(cmd, show_stdout=True, cwd=None,
         env.pop(name, None)
     try:
         proc = subprocess.Popen(
-            cmd, stderr=subprocess.STDOUT, stdin=None, stdout=stdout,
-            cwd=cwd, env=env,
+            cmd, stderr=subprocess.STDOUT, stdin=subprocess.PIPE,
+            stdout=stdout, cwd=cwd, env=env,
         )
+        proc.stdin.close()
     except Exception as exc:
         logger.critical(
             "Error %s while executing command %s", exc, command_desc,

--- a/tests/unit/test_utils.py
+++ b/tests/unit/test_utils.py
@@ -23,8 +23,8 @@ from pip._internal.utils.encoding import auto_decode
 from pip._internal.utils.glibc import check_glibc_version
 from pip._internal.utils.hashes import Hashes, MissingHashes
 from pip._internal.utils.misc import (
-    egg_link_path, ensure_dir, get_installed_distributions, get_prog,
-    normalize_path, rmtree, untar_file, unzip_file,
+    call_subprocess, egg_link_path, ensure_dir, get_installed_distributions,
+    get_prog, normalize_path, rmtree, untar_file, unzip_file,
 )
 from pip._internal.utils.packaging import check_dist_requires_python
 from pip._internal.utils.temp_dir import TempDirectory
@@ -612,3 +612,15 @@ class TestGetProg(object):
             executable
         )
         assert get_prog() == expected
+
+
+def test_call_subprocess_works_okay_when_just_given_nothing():
+    try:
+        call_subprocess([sys.executable, '-c', 'print("Hello")'])
+    except Exception:
+        assert False, "Expected subprocess call to succeed"
+
+
+def test_call_subprocess_closes_stdin():
+    with pytest.raises(InstallationError):
+        call_subprocess([sys.executable, '-c', 'input()'])

